src/coap_client_zephyr.c

```diff
@@ -0,0 +1,1147 @@
+         * fail (during handshake). If this is the case, you can can still connect if you
+         * set CONFIG_GOLIOTH_HOSTNAME_VERIFICATION_SKIP=y, which disables hostname
+         * verification.
+         */
+        ret = zsock_setsockopt(sock, SOL_TLS, TLS_HOSTNAME,
+                       COND_CODE_1(CONFIG_GOLIOTH_HOSTNAME_VERIFICATION_SKIP,
+                           (NULL, 0),
+                           (host, strlen(host) + 1)));
+        if (ret < 0) {
+            return -errno;
+        }
+    }
+
+    if (sizeof(golioth_ciphersuites) > 0) {
+        ret = zsock_setsockopt(sock, SOL_TLS, TLS_CIPHERSUITE_LIST,
+                       golioth_ciphersuites, sizeof(golioth_ciphersuites));
+        if (ret < 0) {
+            return -errno;
+        }
+    }
+
+    return 0;
+}
+
+static int __golioth_send(struct golioth_client* client, uint8_t *data,
+                          size_t len, int flags) {
+    ssize_t sent;
+
+    if (client->sock < 0) {
+        return -ENOTCONN;
+    }
+
+    sent = zsock_send(client->sock, data, len, flags);
+    if (sent < 0) {
+        LOG_ERR("Socket send failed with %d", -errno);
+        return -errno;
+    } else if (sent < len) {
+        return -EIO;
+    }
+
+    return 0;
+}
+
+static int __golioth_send_empty_coap(struct golioth_client* client) {
+    struct coap_packet packet;
+    uint8_t buffer[GOLIOTH_EMPTY_PACKET_LEN];
+    int err;
+
+    err = coap_packet_init(&packet, buffer, sizeof(buffer),
+                   COAP_VERSION_1, COAP_TYPE_NON_CON,
+                   0, NULL,
+                   COAP_CODE_EMPTY, coap_next_id());
+    if (err) {
+        return err;
+    }
+
+    return __golioth_send(client, packet.data, packet.offset, 0);
+}
+
+static int golioth_connect_sockaddr(struct golioth_client* client, const char *host,
+                    struct sockaddr *addr, socklen_t addrlen) {
+    int sock;
+    int ret;
+    int err = 0;
+
+    sock = zsock_socket(addr->sa_family, SOCK_DGRAM, IPPROTO_DTLS_1_2);
+    if (sock < 0) {
+        LOG_ERR("Failed to create socket: %d", -errno);
+        return -errno;
+    }
+
+    err = golioth_setsockopt_dtls(client, sock, host);
+    if (err) {
+        LOG_ERR("Failed to set DTLS socket options: %d", err);
+        goto close_sock;
+    }
+
+    ret = zsock_connect(sock, addr, addrlen);
+    if (ret < 0) {
+        err = -errno;
+        LOG_ERR("Failed to connect to socket: %d", err);
+        goto close_sock;
+    }
+
+    client->sock = sock;
+
+    /* Send empty packet to start TLS handshake */
+    err = __golioth_send_empty_coap(client);
+    if (err) {
+        LOG_ERR("Failed to send empty CoAP message: %d", err);
+        client->sock = -1;
+    }
+
+close_sock:
+    if (err) {
+        __golioth_close(sock);
+    }
+
+    return err;
+}
+
+#if CONFIG_GOLIOTH_LOG_LEVEL >= LOG_LEVEL_DBG
+#define LOG_SOCKADDR(fmt, addr)                         \
```

Comment:
Does this need to be a macro, or can we make it a function?

Reply:
`fmt` needs to be evaluated at compile time in invocation of `LOG_DBG(fmt, ...);`, so that is the main reason why it is a macro.